### ISSUE TITLE
PJRT `xm.get_ordinal` and `xm.all_gather` fails in backward pass due to thread locals not accessible from autograd

### ISSUE THREAD

**Issue by ronghanghu** [pytorch/xla#3949]
## 🐛 Bug

https://github.com/pytorch/xla/pull/3902 addressed the issue of mismatched ranks in PJRT collective ops and made `xm.get_ordinal` return the corresponding ranks used by collective ops such as `xm.reduce_scatter`, which resolves many PJRT use cases.

However, further tests show that the thread locals `pjrt._PJRT_ORDINALS` (which `xm.get_ordinal` and consequently `xm.all_gather` are building upon) can only be read correctly during the forward pass, but cannot be read during the backward pass when invoked by PyTorch's autograd engine.

This breaks several use cases where one needs to call `xm.get_ordinal` or `xm.all_gather` in the autograd backward pass, such as in a custom-defined `torch.autograd.Function` or in more complicated backward hooks such as FSDP or other cases.

(This issue was discovered when testing FSDP under PJRT, which involves calling `xm.all_gather` from a backward hook of a network module.)

## To Reproduce

The example below contains a simple test case with a PyTorch autograd function `MyCustomFunction` that defines a dummy forward and backward method. It shows that `xm.get_ordinal` always returns 0 when invoked in the backward pass. This is because the thread locals `pjrt._PJRT_ORDINALS` cannot be accessed from the autograd engine that calls the backward functions or hooks.

1. Allocate a new v3-8 TPU VM with `tpu-vm-pt-1.12` runtime and install the nightly PyTorch/XLA wheels that contain https://github.com/pytorch/xla/pull/3704
```
# torch, torchvision and torch_xla
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch-nightly+20220829-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torchvision-nightly+20220829-cp38-cp38-linux_x86_64.whl
sudo pip3 install https://storage.googleapis.com/tpu-pytorch/wheels/tpuvm/torch_xla-nightly+20220829-cp38-cp38-linux_x86_64.whl

# libtpu
sudo pip3 install https://storage.googleapis.com/cloud-tpu-tpuvm-artifacts/wheels/libtpu-nightly/libtpu_nightly-0.1.dev20220623-py3-none-any.whl
```
2. Save the following content to a file `test_pjrt_thread_locals_in_backward.py`
```python3
import torch
import torch_xla.core.xla_model as xm
from torch_xla.experimental import pjrt


class MyCustomFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        rank = xm.get_ordinal()
        pjrt_ordinal = getattr(pjrt._PJRT_ORDINALS, "global_ordinal", None)
        ctx.forward_rank = rank
        print(f"forward rank: {rank}\n" f"\tforward pjrt._PJRT_ORDINALS.global_ordinal: {pjrt_ordinal}\n", end="")
        return x

    @staticmethod
    def backward(ctx, grad_output):
        rank = xm.get_ordinal()
        pjrt_ordinal = getattr(pjrt._PJRT_ORDINALS, "global_ordinal", None)
        forward_rank = ctx.forward_rank
        print(
            f"backward rank: {rank} vs forward rank: {forward_rank}\n"
            f"\tbackward pjrt._PJRT_ORDINALS.global_ordinal: {pjrt_ordinal}\n",
            end="",
        )

        return grad_output


def test_pjrt_thread_locals():
    x = torch.ones(1, requires_grad=True, device=xm.xla_device())
    y = MyCustomFunction.apply(x)
    y.backward()
    xm.mark_step()


if __name__ == "__main__":
    pjrt.run_multiprocess(test_pjrt_thread_locals)
```
3. Run this file under PJRT via `PJRT_DEVICE=TPU python3 test_pjrt_thread_locals_in_backward.py`, which shows that the during the backward pass `xm.get_ordinal()` always returns 0 and the PJRT thread local `pjrt._PJRT_ORDINALS` does not contain the `global_ordinal` attribute, printing the following
```
forward rank: 0
        forward pjrt._PJRT_ORDINALS.global_ordinal: 0
forward rank: 1
        forward pjrt._PJRT_ORDINALS.global_ordinal: 1
forward rank: 2
        forward pjrt._PJRT_ORDINALS.global_ordinal: 2
forward rank: 3
        forward pjrt._PJRT_ORDINALS.global_ordinal: 3
forward rank: 4
        forward pjrt._PJRT_ORDINALS.global_ordinal: 4
forward rank: 5
        forward pjrt._PJRT_ORDINALS.global_ordinal: 5
forward rank: 6
        forward pjrt._PJRT_ORDINALS.global_ordinal: 6
forward rank: 7
        forward pjrt._PJRT_ORDINALS.global_ordinal: 7
backward rank: 0 vs forward rank: 0
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 1
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 2
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 3
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 4
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 5
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 6
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
backward rank: 0 vs forward rank: 7
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
```

As can be seen above, `xm.get_ordinal` always returns 0 when invoked in the backward pass (inside the autograd engine). Also, the thread locals `pjrt._PJRT_ORDINALS` is empty in the backward pass and does not contain `global_ordinal`.

**A consequence to this is that `xm.all_gather` won't work correctly when invoked in the backward pass under PJRT**, since it depends on `xm.get_ordinal`. Under `pin_layout=True` (by default), the API `xm.all_gather` doesn't actually perform all-gather, but instead, it first pads the inputs with zeros and then performs all_reduce as in https://github.com/pytorch/xla/blob/ff2e0ea559b5f8785c7f577834af2b5af84b6849/torch_xla/core/xla_model.py#L673-L678

## Expected behavior

The API `xm.get_ordinal` should be able to return the correct rank in both forward pass and backward pass.

And `xm.all_gather` under `pin_layout=True` (by default) should work in both forward pass and backward pass.

## Environment

 - Reproducible on XLA backend [CPU/TPU]: v3-8 TPU VM with `tpu-vm-pt-1.12` runtime
 - torch_xla version: nightly 20220829 (see above)

cc: @will-cromar @JackCaoG

**Comment by ronghanghu** [comment]
Since we are aware of this issue now, we can try to avoid it by not calling `xm.get_ordinal()` from backward functions or hooks (and instead saving them from the forward pass), and also patch `xm.all_gather` (which is actually implemented via padding + all-reduce) to have the correct rank if we need to use it from the backward pass. So this issue is not blocking us for now.

But overall it still looks like a bug and might be a surprise to users, so I feel it would be great to have a solution for it :)

**Comment by will-cromar** [comment]
Thanks again @ronghanghu for a detailed bug report!

Interestingly, `xla_device()` also relies on thread-local storage, specifically this variable:

https://github.com/pytorch/xla/blob/dbc67604cfff305a37544513c20e377b254054ef/torch_xla/csrc/device.cpp#L12

But, for some reason, the device itself seems to be correctly handled in the backward pass. I applied this patch to your test case:

```
        device = str(xm.xla_device())
        pjrt_ordinal = getattr(pjrt._PJRT_ORDINALS, "global_ordinal", None)
        forward_rank = ctx.forward_rank
        print(
            f"backward rank: {rank} vs forward rank: {forward_rank}\n"
            f"\tbackward pjrt._PJRT_ORDINALS.global_ordinal: {pjrt_ordinal}\n"
            f"\tdevice: {device}\n",
            end="",
        )
```

I end up with this output if I run 1 process with 4 threads on PJRT `PJRT_DEVICE=TPU TPU_PROCESS_BOUNDS=1,1,1 TPU_VISIBLE_DEVICES=0,1,2,3 python ./autograd.py`

```
backward rank: 0 vs forward rank: 1
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
        device: xla:1
backward rank: 0 vs forward rank: 0
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
        device: xla:0
backward rank: 0 vs forward rank: 2
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
        device: xla:2
backward rank: 0 vs forward rank: 3
        backward pjrt._PJRT_ORDINALS.global_ordinal: None
        device: xla:3
```

I wonder if there's special handling for this variable we can replicate for the ordinals. Or, maybe `thread_local` C++ variables are handled differently by autograd. I'm hoping the latter is true, because then there would be a straightforward potential fix by moving these ordinals into C++.

**Comment by ronghanghu** [comment]
@will-cromar Thanks for the follow-up and tests on `xm.xla_device()`. I guess the latter (that thread_local C++ variables are handled differently by autograd) is more likely to be the case as it's my understanding that PyTorch autograd engine is above PyTorch/XLA tracing (and other XLA details) and it shouldn't have special handling for this `g_current_device` variable.

**Comment by ronghanghu** [comment]
I tried out the nightly 20220914 wheels and confirm that this issue is now resolved in our use cases.
